owncloudData/src/main/java/com/owncloud/android/data/providers/LocalStorageProvider.kt

```diff
@@ -211,6 +212,16 @@ sealed class LocalStorageProvider(private val rootFolderName: String) {
         return fileToDelete.deleteRecursively()
     }
 
+    fun deleteLocalFolderIfItHasNoFilesInside(ocFolder: OCFile) {
+        val safeStoragePath = ocFolder.getStoragePathOrExpectedPathForFile()
+        val folder = File(safeStoragePath)
+
+        val filesInFolder = folder.listFiles()
+        if (filesInFolder.isNullOrEmpty()) {
+            folder.deleteRecursively()
+        }
```

Comment:
Calling `folder.delete()` directly could be fine.

Listing files could take some time of the folder contain too many files. We just need to know where the folder contains no files, or at least one; getting the whole list is irrelevant.
The `deleteRecursively` will also have some little overhead when we know that the folder is empty (because we've just checked it).

By calling directly the `delete()` method, we don't need to list the whole folder because the method will fail if the folder isn't empty. We might need to ensure that the `ocFolder` is actually a folder so we don't try to remove a file (the check might still be needed regardless of the proposed changes, unless there are guarantees that the `ocFolder` will NEVER be a file)

Reply:
This method will only be invoked for folders. You can see that its only use in `OCFileRepository` is surrounded by `if (file.isFolder) { }`.

OTOH, I'm not sure how the `delete()` method handles the failure if we try to remove a non-empty file. Looking at the documentation, it seems that it doesn't throw an exception because of this. The only signal we have is that it may return `false`, but documentation only specifies that it returns `false` if the folder couldn't be deleted correctly, could be that it returns false and not due to the folder not being empty (for example, due to permissions or whatever). So, I don't really know if using `delete()` is a good approach in this case

Reply:
> This method will only be invoked for folders. You can see that its only use in OCFileRepository is surrounded by if (file.isFolder) { }.

But this method is public and it could be called from anywhere. Furthermore, specific requirements to call this method, such as `ocFolder` being an actual folder and not a file, aren't documented, so there is a risk that new code will use this method incorrectly and will cause issues.

> but documentation only specifies that it returns false if the folder couldn't be deleted correctly, could be that it returns false and not due to the folder not being empty (for example, due to permissions or whatever). 

I haven't checked the docs for the `deleteRecursively` method, but it should have the same problem. The question is whether we can and / or need to distinguish any of the error cases. With the current code, it doesn't seem we're interested in the reason of the failure.

Assuming we have only 2 error conditions (folder not empty and permission issue), we could assume that everything is fine and try to delete the folder via `delete()`, then if the method fails we could check if the folder is empty in order to rule out which error case happened. This should perform faster assuming the success case is more predominant.

Reply:
We've looked at the Java documentation and the `delete()` method doesn't specify when the method receives a false or an exception. So to try to cover at least the case of deleting the folder when it's empty, we're going to keep this part of the code:
```
val filesInFolder = folder.listFiles()
        if (filesInFolder.isNullOrEmpty()) {
```
 and change the `folder.deleteRecursively()` to the `delete()`.